Handle headers that are defined more than once

mbland opened the pull request:
Per @jehiah's suggestion in bitly/oauth2_proxy#147.

Also ensures that the path, query, and fragment of the URL is factored in.

@jehiah, to your other points from bitly/oauth2_proxy#147 (and I'd like to hear others weigh in as well):

I've tried using an integer index instead of the full header name to distinguish header values that correspond to the same header. How does that sound as an alternative to including the full header name?

Not sure if enforcing that the header fields are processed in sorted order is useful at this level of abstraction. The application could certainly enforce this policy, though. Thoughts?

cc: @jcscottiii as my primary Go reviewer.

cc: @dhcole, @jmcarp and @ozzyjohnson for the general approach. I'll port whatever happens in this PR to https://rubygems.org/gems/hmac_authentication and https://www.npmjs.com/package/hmac-authentication, and I'll also get a Python module started shortly.

mbland commented:
Another alternative to the numeric prefixes is to concatenate multiple headers into a comma-separated string. Thoughts?

mbland commented:
In fact, I think this will play more nicely across language implementations, based on my experience with [the Node HTTP request implementation](https://nodejs.org/api/http.html#http_http):

> Defined headers that allow multiple values are concatenated with a , character, except for the set-cookie and cookie headers which are represented as an array of values. Headers such as content-length which can only have a single value are parsed accordingly, and only a single value is represented on the parsed object.

Going to implement this now.

mbland commented:
BTW, while updating 18F/hmac_authentication_gem, I learned that the [Ruby Net::HTTPHeader class only honors the last header defined with the same name](https://github.com/rubysl/rubysl-net-http/blob/2.0/lib/net/http/header.rb#L15). That's something to shout off of the rooftops.

mbland commented:
Gang, I've made a bunch of refinements; if it's too much, let me know and I'll start breaking some of them out. (But since it's such a small library, I'm hoping it's not that bad...)

Any chance anyone'll have cycles to review today or tomorrow?

jcscottiii commented:
@mbland Taking a look now.

jcscottiii commented on hmacauth.go:
```diff
@@ -5,35 +5,100 @@ import (
 	algorithmName = make(map[crypto.Hash]string)
 	for name, algorithm := range supportedAlgorithms {
 		algorithmName[algorithm] = name
+		if algorithm.Available() == false {
```
Looking at the documentation for Available() . It says `Available reports whether the given hash function is linked into the binary.`
When would one of those algorithms NOT be linked into the binary? Just a comment to explain.

mzia replied:
My understanding is that IF your code is linking to a version of a crypto library when building this service/library (as HASH implementations grow over time), you want to verify against the requestors use of a HASH function so that you can perform the same function to reproduce the HASH using algors such as SHA1, SHA2, SHA3, whereby you may not have SHA3 implementation support yet.

mbland replied:
Done.

jcscottiii commented on hmacauth.go:
```diff
@@ -5,35 +5,100 @@ import (
 
-func StringToSign(req *http.Request, headers []string) string {
+func DigestNameToCryptoHash(name string) (result crypto.Hash, err error) {
+	if result = supportedAlgorithms[name]; result == crypto.Hash(0) {
```
Use the map key test talked about [here](https://blog.golang.org/go-maps-in-action)

``` go
if result, supported := supportedAlgorithms[name]; !supported {
...
```

mbland replied:
Done.

jcscottiii commented on hmacauth.go:
```diff
@@ -5,35 +5,100 @@ import (
+}
+
+func CryptoHashToDigestName(id crypto.Hash) (result string, err error) {
+	if result = algorithmName[id]; result == "" {
```
change to similar map key test mentioned above

mbland replied:
Done.

jcscottiii commented on hmacauth.go:
```diff
@@ -5,35 +5,100 @@ import (
+	return
+}
+
+type HmacAuth struct {
```
change name of struct from `HmacAuth` to `hmacAuth`. This will ensure hmacAuth can only be constructed via your `NewHmacAuth` method.

mbland replied:
So I did this, but then needed to export a new HmacAuth interface to ensure callers had access to the methods. I also switched the test namespace to `hmacauth_test`, per the advice from the ginkgo docs (even though I'm not using ginkgo here).

jcscottiii commented on hmacauth.go:
```diff
@@ -5,35 +5,100 @@ import (
+	headers []string) *HmacAuth {
+	if hash.Available() == false {
+		var name string
+		if name = algorithmName[hash]; name == "" {
```
similar map key test here.

mbland replied:
Done.

jcscottiii commented on hmacauth.go:
```diff
@@ -5,35 +5,100 @@ import (
+		panic("hmacauth: hash algorithm " + name + " is unavailable")
+	}
+	canonicalHeaders := make([]string, len(headers))
+	for i, h := range headers {
```
`h` -> `header`

jcscottiii commented:
@mbland finished! It looks good!
In addition, you should run `golint` on your [code](https://github.com/golang/lint) and fix those errors.
There's no standard yet to use this tool yet, but you can try [gometalinter](https://github.com/alecthomas/gometalinter) if you want. I'm still evaluating it myself.

jcscottiii commented:
also you should put that assert library in the Go testing cookbook! really clean.

mbland commented:
@jcscottiii Addressed all your comments, including taking a pass with `golint` and `gometalinter` (which, ironically, required me to undo the Ginkgo-esque `import . "github.com/18F/hmacauth"` directive. If it looks good to you, would you mind merging?

jcscottiii commented:
@mbland merged! :)

mbland commented:
Thanks, @jcscottiii!
